jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -268,7 +268,134 @@ private static StringBuilder encodePath(StringBuilder buf, String path, int offs
 
         return buf;
     }
-    
+
+    /**
+     * Encode a raw URI String and convert any raw spaces to
+     * their "%20" equivalent.
+     *
+     * @param str input raw string
+     * @return output with spaces converted to "%20"
+     */
+    public static String encodeSpaces(String str)
+    {
+        return StringUtil.replace(str, " ", "%20");
+    }
+
+    /**
+     * Encode a raw String and convert any specific characters to their URI encoded equivalent.
+     *
+     * @param str input raw string
+     * @param charsToEncode the list of raw characters that need to be encoded (if encountered)
+     * @return output with specified characters encoded.
+     */
+    @SuppressWarnings("Duplicates")
+    public static String encodeSpecific(String str, String charsToEncode)
+    {
+        if ((str == null) || (str.length() == 0))
+            return null;
+
+        if ((charsToEncode == null) || (charsToEncode.length() == 0))
+            return str;
+
+        char[] find = charsToEncode.toCharArray();
+        int len = str.length();
+        StringBuilder ret = new StringBuilder((int)(len * 0.20d));
+        for (int i = 0; i < len; i++)
+        {
+            char c = str.charAt(i);
+            boolean escaped = false;
+            for (char f : find)
+            {
+                if (c == f)
+                {
+                    escaped = true;
+                    ret.append('%');
+                    int d = 0xf & ((0xF0 & c) >> 4);
+                    ret.append((char)((d > 9 ? ('A' - 10) : '0') + d));
+                    d = 0xf & c;
+                    ret.append((char)((d > 9 ? ('A' - 10) : '0') + d));
+                    break;
+                }
+            }
+            if (!escaped)
+            {
+                ret.append(c);
+            }
+        }
+        return ret.toString();
+    }
+
+    /**
+     * Decode a raw String and convert any specific URI encoded sequences into characters.
+     *
+     * @param str input raw string
+     * @param charsToDecode the list of raw characters that need to be decoded (if encountered), leaving all other encoded sequences alone.
+     * @return output with specified characters decoded.
+     */
+    @SuppressWarnings("Duplicates")
+    public static String decodeSpecific(String str, String charsToDecode)
```

Comment:
I don't think decode specific makes sense.  I think it is always decode all.  In fact we could almost always encode all as well, with the only exceptions being if we encode chars like : and / which are part of the URI?

Reply:
We would need to know if the Mongo specific behavior would like or dislike this complete approach.

Reply:
encode/decode specific is typically for places where special characters cause problems.
and where blanket encode/decode can cause even greater problems. (filesystem refs, special characters that must be included, etc)
Keep in mind that these encode/decode specific methods are used in places where we need to have a symmetric encode/decode of String information, but protected ourselves from known special characters.

Reply:
OK, I'm sold on encode specific... but I'm dubious about the benefit of decoding specific characters only.     If the string has %XY and XY represents a character that should not be decoded, then that is an encoding error and the % should have been encoded.   So I would strongly prefer to not have the decode specific (specially as it is not used).